### ISSUE TITLE
proposal: warn when trying to catch KeyError in defaultdict lookup

### ISSUE THREAD

**Issue by jgehrcke** [pylint-dev/pylint#8666]
### Current problem

I just debugged a funny situation where I did
```python
    try:
        matching_results = bmrt_cache["by_benchmark_name"][bname]
    except KeyError:
        return f"benchmark name not known: `{bname}`"
```
In this snippet, `bmrt_cache["by_benchmark_name"]` is a `defaultdict(list)`. That is, this never actually raises a `KeyError`, but instead inserts the key. Wwhich was an unintended side effect in this case.

Of course this was a little stupid and such, but yeah that's how bugs happen. Just for completeness: we use mypy and pylint and flake8 on this code base, there was no warning.

### Desired solution

If that's not possible to detect with reasonable confidence then that's OK of course, this might be pretty tricky. But yeah I thought that it's worth asking: can a tool like `pylint` detect this situation, with certainty? Can we even emit an error in this case?

If not with certainty: maybe we can emit a warning for the `except KeyError` line, stating that it might never be reachable?

**Comment by Pierre-Sassoulas** [comment]
Thank you for opening the issue, it seems reasonable.

Detecting with certainty or not depend on the scope where the default dict is declared. 

This will be uninferable for example:
```python
    def function(bmrt_cache, bname):
        try:
            matching_results = bmrt_cache["by_benchmark_name"][bname]
        except KeyError:
            return f"benchmark name not known: `{bname}`"
```            
But if we have more info we're going to be able to infer the value (modulo conditional and hard to understand code construct, or code that is "too dynamic").

**Comment by jgehrcke** [comment]
Thank you for the initial feedback on this @Pierre-Sassoulas and a huge thank you for your work on pylint. I am glad I asked!

**Comment by jgehrcke** [comment]
@Pierre-Sassoulas I wonder if this is worth exploring further. I myself feel rather clueless with respect to contributing here however. Is this easy? :)

**Comment by Pierre-Sassoulas** [comment]
Modifying inference in astroid is not very easy, I wouldn't start my first contribution with this. But there's some documentation for it here: https://pylint.readthedocs.io/projects/astroid/en/latest/inference.html
